File: solutions/Leetcode_1209/leetcode_1209.cc
#include <iosteam>
#include <string>
#include <vector>

using namespace std;

class Solution1
{
public:
    string removeDuplicates(string s, int k)
    {
        vector<pair<char, int>> stk;
        for (auto x : s)
        {
            if (stk.empty() || stk.back().first != x)
            {
                stk.push_back({x, 1});
            }
            else if (stk.back().second + 1 == k)
            {
                stk.pop_back();
            }
            else
            {
                ++stk.back().second;
            }
        }
        string ans;
        for (auto p : stk)
            ans.append(p.second, p.first);
        return ans;
    }
};

int main()
{
    string s = "abcd";
    int k = 3;
    string ans = Solution1().removeDuplicates(s, k);
    cout << ans << endl;
    return 0;
}

File: solutions/Leetcode_1209/leetcode_1209.py
class Solution1:
    def remove_duplicates(self, s: str, k: int) -> str:
        if len(s) < k:
            return s
        ans = []
        for x in s:
            if not ans or x != ans[-1][0]:
                ans.append([x, 1])
            elif ans[-1][1] + 1 == k:
                ans.pop()
            else:
                ans[-1][1] += 1
        ret = ''.join([char * count for char, count in ans])
        return ret


if __name__ == '__main__':
    s = 'abcd'
    k = 2
    ans = Solution1().remove_duplicates(s, k)
    print(ans)
